File: cpp/arrays/frequency012.cpp
#include <bits/stdc++.h>
using namespace std;

void printArray(int arr[], int n){
    for(int i=0; i<n; i++)
        cout << arr[i] << " ";
}

int swap(int *a, int *b){
    *a = *a ^ *b;
    *b = *a ^ *b;
    *a = *a ^ *b;
}

int frequency1(int arr[], int n){
    // int sortedArr[n];
    int count0 = 0, count1 = 0, count2 = 0;
    for(int i=0; i<n; i++){
        if(arr[i]==0)
            count0++;
        else if(arr[i]==1)
            count1++;
        else
            count2++;
    }

    for(int i=0; i<count0; i++)
        arr[i] = 0;
    for(int i=count0; i<count0+count1; i++)
        arr[i] =1;
    for(int i=count0+count1; i<n; i++)
        arr[i] = 2;

    printArray(arr, n);
}

int dutchNationalFlagAlgo(int arr[], int n) {
    int low = 0, mid = 0, high = n-1;
    while(mid <= high) {
        if(arr[mid]==0){
            swap(arr+mid, arr+low);
            low++;
            mid++;
        } else if(arr[mid]==1){
            mid++;
        } else {
            swap(arr+mid, arr+high);
            high--;
        }
    }

    printArray(arr, n);
}

int main(){
    int arr[] = {0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1};
    int n = sizeof(arr)/ sizeof(arr[0]);
    cout << "Array: ";
    printArray(arr, n);
    cout << "\nSorted array: ";
    // frequency1(arr, n);
    dutchNationalFlagAlgo(arr, n);
}

File: Basic-Data-Structures/frequency012.py
def frequency1(arr):
    n = len(arr)
    count = [0, 0, 0]

    for key in arr:
        count[key] += 1

    for val in range(count[0]):
        arr[val] = 0
    
    for val in range(count[0], count[1]+count[0]):
        arr[val] = 1
    
    for val in range(count[1]+count[0], n):
        arr[val] = 2

    return arr

def frequency2(arr):
    index = 0
    sortedArr = []

    for key in arr:
        if key == 2:
            sortedArr.append(key)
        elif key == 1:
            sortedArr.insert(index, key)
            index += 1
        else:
            sortedArr.insert(0, key)
            index += 1

    return sortedArr

def dutchNationalFlagAlgo(arr):
    low = 0
    mid = 0
    high = len(arr) - 1

    while mid <= high:
        if arr[mid] == 0:
            arr[mid], arr[low] = arr[low], arr[mid]
            mid += 1
            low += 1
        elif arr[mid] == 1:
            mid += 1
        else:
            arr[mid], arr[high] = arr[high], arr[mid]
            high -= 1
    return arr

arr = [0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1]
# print(frequency1(arr))
print(frequency2(arr))
# print(dutchNationalFlagAlgo(arr))